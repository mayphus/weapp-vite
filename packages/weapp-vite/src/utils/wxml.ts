import type { Buffer } from 'node:buffer'
import { Parser } from 'htmlparser2'
import MagicString from 'magic-string'

// https://developers.weixin.qq.com/miniprogram/dev/framework/view/wxml/event.html
export function processWxml(wxml: string | Buffer) {
  const ms = new MagicString(wxml.toString())
  // transformOn
  // https://github.com/vuejs/core/blob/76c43c6040518c93b41f60a28b224f967c007fdf/packages/compiler-core/src/transforms/vOn.ts
  const parser = new Parser({
    onattribute(name, _value, _quote) {
      // 事件绑定
      if (name.startsWith('@')) {
        const start = parser.startIndex
        const end = parser.startIndex + name.length
        const { dir, mods } = name.split('.')
          .filter(x => x)
          .reduce<{ dir: string, mods: Record<string, boolean> }>((acc, cur, idx) => {
            if (idx === 0) {
              acc.dir = cur
            }
            else {
              acc.mods[cur] = true
            }
            return acc
          }, { dir: '', mods: {} })

        let rep: string
        if (mods.catch && mods.capture) {
          rep = `capture-catch:${dir.slice(1)}`
        }
        else if (mods.catch) {
          rep = `catch:${dir.slice(1)}`
        }
        else if (mods.mut) {
          rep = `mut-bind:${dir.slice(1)}`
        }
        else if (mods.capture) {
          rep = `capture-bind:${dir.slice(1)}`
        }
        else {
          rep = `bind:${dir.slice(1)}`
        }
        if (rep) {
          ms.update(start, end, rep)
        }
      }
    },
  })
  parser.write(
    ms.original,
  )
  parser.end()
  return {
    code: ms.toString(),
  }
}
