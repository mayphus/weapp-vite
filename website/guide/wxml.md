# WXML 增强

## 事件绑定增强

默认情况下，启用 `weapp.enhance.wxml` 模式，可以使用类似 `vue` 的事件绑定写法:

```html
<!-- 源代码 -->
<view @tap="hello"></view>
<!-- 编译结果 -->
<view bind:tap="hello"></view>
```

以下表格为 `tap` 事件的一一对应，其他事件同理:

| 源代码                                      | 编译结果            |
| ------------------------------------------- | ------------------- |
| `@tap`                                      | `bind:tap`          |
| `@tap.catch`                                | `catch:tap`         |
| `@tap.mut`                                  | `mut-bind:tap`      |
| `@tap.capture`                              | `capture-bind:tap`  |
| `@tap.capture.catch` / `@tap.catch.capture` | `capture-catch:tap` |

更多事件请查看 [事件分类](https://developers.weixin.qq.com/miniprogram/dev/framework/view/wxml/event.html#%E4%BA%8B%E4%BB%B6%E5%88%86%E7%B1%BB)

## 如何关闭

可在 `vite.config.ts` 中通过设置 `weapp.enhance.wxml` 为 `false` 关闭这个增强行为。

关闭后可略微提升性能。
