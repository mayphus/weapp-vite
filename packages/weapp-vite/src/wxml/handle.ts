import type { HandleWxmlOptions } from '../types'
import type { scanWxml } from './scan'
import { defu } from '@weapp-core/shared'
import MagicString from 'magic-string'
import { normalizeWxsFilename, transformWxsCode } from '../wxs'

export function handleWxml(data: ReturnType<typeof scanWxml>, options?: HandleWxmlOptions) {
  const opts = defu<Required<HandleWxmlOptions>, HandleWxmlOptions[]>(options, {
    removeComment: true,
    transformEvent: true,
  })
  const {
    code,
    removeStartStack,
    removeEndStack,
    commentTokens,
    eventTokens,
    inlineWxsTokens,
    removeWxsLangAttrTokens,
    wxsImportNormalizeTokens,
    components,
    deps,
  } = data
  const ms = new MagicString(code)

  for (const { start, end, value } of wxsImportNormalizeTokens) {
    ms.update(start, end, normalizeWxsFilename(value))
  }

  for (const { start, end } of removeWxsLangAttrTokens) {
    ms.update(start, end, '')
  }
  for (const { end, start, value } of inlineWxsTokens) {
    const { result } = transformWxsCode(value)
    if (result?.code) {
      ms.update(start, end, `\n${result.code}`)
    }
  }
  if (opts.transformEvent) {
    for (const { end, start, value } of eventTokens) {
      ms.update(start, end, value)
    }
  }

  for (let i = 0; i < removeStartStack.length; i++) {
    const startIndex = removeStartStack[i]
    for (let j = i; j < removeEndStack.length; j++) {
      const endIndex = removeEndStack[j]
      if (endIndex > startIndex) {
        ms.remove(startIndex, endIndex)
        break
      }
    }
  }
  // remove comments
  if (opts.removeComment) {
    for (const { end, start } of commentTokens) {
      ms.remove(start, end)
    }
  }

  return {
    code: ms.toString(),
    components,
    deps,
  }
}
