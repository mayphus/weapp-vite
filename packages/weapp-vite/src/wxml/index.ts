import type { Buffer } from 'node:buffer'
import { defu } from '@weapp-core/shared'
import { Parser } from 'htmlparser2'
import MagicString from 'magic-string'
import { isBuiltinComponent } from '../auto-import-components/builtin'
import { jsExtensions } from '../constants'
import { normalizeWxsFilename, transformWxsCode } from '../wxs'

const srcImportTagsMap: Record<string, string[]> = {
  // audio: ['src', 'poster'],
  // video: ['src', 'poster'],
  // image: ['src'],
  // https://developers.weixin.qq.com/miniprogram/dev/reference/wxs/01wxs-module.html
  wxs: ['src'],
  // https://developers.weixin.qq.com/miniprogram/dev/reference/wxml/import.html
  // import: ['src'],
  // include: ['src'],
}

export interface WxmlDep {
  tagName: string
  start: number
  end: number
  quote: string | null | undefined
  name: string
  value: string
  attrs: Record<string, string>
}

export interface ScanComponentItem {
  start: number
  end: number
}

export interface ProcessWxmlOptions {
  excludeComponent?: (tagName: string) => boolean
}

// https://developers.weixin.qq.com/miniprogram/dev/framework/view/wxml/event.html
export function processWxml(wxml: string | Buffer, options?: ProcessWxmlOptions) {
  const { excludeComponent } = defu<Required<ProcessWxmlOptions>, ProcessWxmlOptions[]>(options, {
    excludeComponent: (tagName) => {
      return isBuiltinComponent(tagName)
    },
  })
  const ms = new MagicString(wxml.toString())
  const deps: WxmlDep[] = []
  let currentTagName = ''
  let importAttrs: undefined | string[]
  let attrs: Record<string, string> = {}
  const components: Record<string, ScanComponentItem[]> = {}
  let tagStartIndex = -1
  // transformOn
  // https://github.com/vuejs/core/blob/76c43c6040518c93b41f60a28b224f967c007fdf/packages/compiler-core/src/transforms/vOn.ts
  const parser = new Parser(
    {
      onopentagname(name) {
        currentTagName = name
        importAttrs = srcImportTagsMap[currentTagName]
        tagStartIndex = parser.startIndex
      },
      onattribute(name, value, quote) {
        attrs[name] = value
        if (importAttrs) {
          for (const attrName of importAttrs) {
            if (attrName === name) {
              deps.push({
                name,
                value,
                quote,
                tagName: currentTagName,
                start: parser.startIndex,
                end: parser.endIndex,
                attrs,
              })
              if (currentTagName === 'wxs' && name === 'src') {
                if (/\.wxs.[jt]s$/.test(value)) {
                // 5 是 'src="'.length
                // 1 是 '"'.length
                  ms.update(parser.startIndex + 5, parser.endIndex - 1, normalizeWxsFilename(value))
                }
              }
            }
          }
        }
        // 事件绑定
        if (name.startsWith('@')) {
          const start = parser.startIndex
          const end = parser.startIndex + name.length
          const { dir, mods } = name.split('.')
            .filter(x => x)
            .reduce<{ dir: string, mods: Record<string, boolean> }>((acc, cur, idx) => {
              if (idx === 0) {
                acc.dir = cur
              }
              else {
                acc.mods[cur] = true
              }
              return acc
            }, { dir: '', mods: {} })

          let rep: string
          if (mods.catch && mods.capture) {
            rep = `capture-catch:${dir.slice(1)}`
          }
          else if (mods.catch) {
            rep = `catch:${dir.slice(1)}`
          }
          else if (mods.mut) {
            rep = `mut-bind:${dir.slice(1)}`
          }
          else if (mods.capture) {
            rep = `capture-bind:${dir.slice(1)}`
          }
          else {
            rep = `bind:${dir.slice(1)}`
          }
          if (rep) {
            ms.update(start, end, rep)
          }
        }
        // 移除内联 wxs 的 lang
        if (currentTagName === 'wxs' && name === 'lang' && jsExtensions.includes(value)) {
          ms.update(parser.startIndex, parser.endIndex, '')
        }
      },
      onclosetag() {
        if (!excludeComponent(currentTagName)) {
          if (Array.isArray(components[currentTagName])) {
            components[currentTagName].push({
              start: tagStartIndex,
              end: parser.endIndex + 1,
            })
          }
          else {
            components[currentTagName] = [{
              start: tagStartIndex,
              end: parser.endIndex + 1,
            }]
          }
        }

        currentTagName = ''
        attrs = {}
        importAttrs = undefined
        tagStartIndex = -1
      },
      ontext(data) {
        if (currentTagName === 'wxs' && jsExtensions.includes(attrs.lang)) {
        // data
          const res = transformWxsCode(data)
          if (res && res.code) {
            ms.update(parser.startIndex, parser.endIndex, `\n${res.code}`)
          }
        }
      },
    },
    {
      lowerCaseTags: false,
    },
  )
  parser.write(
    ms.original,
  )
  parser.end()
  return {
    components,
    deps,
    code: ms.toString(),
  }
}
