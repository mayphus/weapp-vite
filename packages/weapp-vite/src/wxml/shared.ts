export const srcImportTagsMap: Record<string, string[]> = {
  // audio: ['src', 'poster'],
  // video: ['src', 'poster'],
  // image: ['src'],
  // https://developers.weixin.qq.com/miniprogram/dev/reference/wxs/01wxs-module.html
  wxs: ['src'],
  // https://developers.weixin.qq.com/miniprogram/dev/reference/wxml/import.html
  import: ['src'],
  include: ['src'],
}

export interface Token {
  start: number
  end: number
  value: string
}
