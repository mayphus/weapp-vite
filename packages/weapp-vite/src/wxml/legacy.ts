import type { Buffer } from 'node:buffer'
import type { ComponentsMap, ProcessWxmlOptions, WxmlDep } from '../types'
import { defu } from '@weapp-core/shared'
import { Parser } from 'htmlparser2'
import MagicString from 'magic-string'
import { isBuiltinComponent } from '../auto-import-components/builtin'
import { jsExtensions } from '../constants'
import { normalizeWxsFilename, transformWxsCode } from '../wxs'
import { srcImportTagsMap } from './shared'

// https://github.com/fb55/htmlparser2/issues/1541
// https://developers.weixin.qq.com/miniprogram/dev/framework/view/wxml/event.html
export function processWxml(wxml: string | Buffer, options?: ProcessWxmlOptions) {
  const opts = defu<Required<ProcessWxmlOptions>, ProcessWxmlOptions[]>(options, {
    excludeComponent: (tagName) => {
      return isBuiltinComponent(tagName)
    },
    platform: 'weapp',
    removeComment: true,
  })
  const ms = new MagicString(wxml.toString())
  const deps: WxmlDep[] = []
  let currentTagName = ''
  let importAttrs: undefined | string[]
  let attrs: Record<string, string> = {}
  const components: ComponentsMap = {}
  let tagStartIndex = 0
  // transformOn
  // https://github.com/vuejs/core/blob/76c43c6040518c93b41f60a28b224f967c007fdf/packages/compiler-core/src/transforms/vOn.ts

  const removeStartStack: number[] = []
  const removeEndStack: number[] = []
  const commentsPositions: { start: number, end: number }[] = []
  const parser = new Parser(
    {
      onopentagname(name) {
        currentTagName = name
        importAttrs = srcImportTagsMap[currentTagName]
        tagStartIndex = parser.startIndex
      },
      onattribute(name, value, quote) {
        attrs[name] = value
        if (importAttrs) {
          for (const attrName of importAttrs) {
            if (attrName === name) {
              deps.push({
                name,
                value,
                quote,
                tagName: currentTagName,
                start: parser.startIndex,
                end: parser.endIndex,
                attrs,
              })
              if (currentTagName === 'wxs' && name === 'src') {
                if (/\.wxs.[jt]s$/.test(value)) {
                  // 5 是 'src="'.length
                  // 1 是 '"'.length
                  ms.update(parser.startIndex + 5, parser.endIndex - 1, normalizeWxsFilename(value))
                }
              }
            }
          }
        }
        // 事件绑定
        if (name.startsWith('@')) {
          const start = parser.startIndex
          const end = parser.startIndex + name.length
          const { dir, mods } = name.split('.')
            .filter(x => x)
            .reduce<{ dir: string, mods: Record<string, boolean> }>((acc, cur, idx) => {
              if (idx === 0) {
                acc.dir = cur
              }
              else {
                acc.mods[cur] = true
              }
              return acc
            }, { dir: '', mods: {} })

          let rep: string
          if (mods.catch && mods.capture) {
            rep = `capture-catch:${dir.slice(1)}`
          }
          else if (mods.catch) {
            rep = `catch:${dir.slice(1)}`
          }
          else if (mods.mut) {
            rep = `mut-bind:${dir.slice(1)}`
          }
          else if (mods.capture) {
            rep = `capture-bind:${dir.slice(1)}`
          }
          else {
            rep = `bind:${dir.slice(1)}`
          }
          if (rep) {
            ms.update(start, end, rep)
          }
        }
        // 移除内联 wxs 的 lang
        if (currentTagName === 'wxs' && name === 'lang' && jsExtensions.includes(value)) {
          ms.update(parser.startIndex, parser.endIndex, '')
        }
      },
      onclosetag() {
        if (currentTagName && !opts.excludeComponent(currentTagName)) {
          if (Array.isArray(components[currentTagName])) {
            components[currentTagName].push({
              start: tagStartIndex,
              end: parser.endIndex + 1,
            })
          }
          else {
            components[currentTagName] = [{
              start: tagStartIndex,
              end: parser.endIndex + 1,
            }]
          }
        }

        currentTagName = ''
        attrs = {}
        importAttrs = undefined
        tagStartIndex = 0
      },
      ontext(data) {
        if (currentTagName === 'wxs' && jsExtensions.includes(attrs.lang)) {
          // data
          const res = transformWxsCode(data)
          if (res?.code) {
            ms.update(parser.startIndex, parser.endIndex, `\n${res.code}`)
          }
        }
      },
      // <!--  #ifdef  %PLATFORM% -->
      // 平台特有的组件
      // <!--  #endif -->
      oncomment(data) {
        let match = /#ifdef\s+(\w+)/.exec(data)
        if (match) {
          if (match[1] !== opts.platform) {
            removeStartStack.push(parser.startIndex)
          }
        }
        match = /#endif/.exec(data)
        if (match) {
          removeEndStack.push(parser.endIndex + 1)
        }
        commentsPositions.push({
          start: parser.startIndex,
          end: parser.endIndex + 1,
        })
      },
    },
    {
      lowerCaseTags: false,
      xmlMode: true,
    },
  )
  parser.write(
    ms.original,
  )
  parser.end()

  for (let i = 0; i < removeStartStack.length; i++) {
    const startIndex = removeStartStack[i]
    for (let j = i; j < removeEndStack.length; j++) {
      const endIndex = removeEndStack[j]
      if (endIndex > startIndex) {
        ms.remove(startIndex, endIndex)
        break
      }
    }
  }
  // remove comments
  if (opts.removeComment) {
    for (const { end, start } of commentsPositions) {
      ms.remove(start, end)
    }
  }

  return {
    components,
    deps,
    code: ms.toString(),
    removeStartStack,
    removeEndStack,
  }
}
