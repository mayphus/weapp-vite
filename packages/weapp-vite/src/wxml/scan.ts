import type { ComponentsMap, ScanWxmlOptions, WxmlDep } from '@/types'
import type { Buffer } from 'node:buffer'
import type { Token } from './shared'
import { isBuiltinComponent } from '@/auto-import-components/builtin'
import { jsExtensions } from '@/constants'
import { defu, objectHash } from '@weapp-core/shared'
import { Parser } from 'htmlparser2'
import { LRUCache } from 'lru-cache'
import { srcImportTagsMap } from './shared'

export interface WxmlToken {
  components: ComponentsMap
  deps: WxmlDep[]
  removeStartStack: number[]
  removeEndStack: number[]
  commentTokens: Token[]
  inlineWxsTokens: Token[]
  wxsImportNormalizeTokens: Token[]
  removeWxsLangAttrTokens: Token[]
  eventTokens: Token[]
  code: string
}

export const scanWxmlCache = new LRUCache<string, WxmlToken>(
  {
    max: 512,
  },
)

function getCacheKey(wxml: string | Buffer, options?: ScanWxmlOptions) {
  return objectHash({
    wxml: wxml.toString(),
    options,
  })
}

export function defaultExcludeComponent(tagName: string) {
  return isBuiltinComponent(tagName)
}

export function scanWxml(wxml: string | Buffer, options?: ScanWxmlOptions) {
  const opts = defu<Required<ScanWxmlOptions>, ScanWxmlOptions[]>(options, {
    excludeComponent: defaultExcludeComponent,
    platform: 'weapp',
  })
  const cacheKey = getCacheKey(wxml, opts)
  const t = scanWxmlCache.get(cacheKey)
  if (t) {
    return t
  }
  const ms = wxml.toString()
  const deps: WxmlDep[] = []
  let currentTagName: string | undefined
  let importAttrs: undefined | string[]
  let attrs: Record<string, string> = {}
  const components: ComponentsMap = {}
  let tagStartIndex = 0
  // transformOn
  // https://github.com/vuejs/core/blob/76c43c6040518c93b41f60a28b224f967c007fdf/packages/compiler-core/src/transforms/vOn.ts

  // 条件编译注释
  const removeStartStack: number[] = []
  const removeEndStack: number[] = []
  // 注释
  const commentTokens: Token[] = []
  // 内联wxs
  const inlineWxsTokens: Token[] = []
  // 处理 xxx.wxs.ts 转变为合法引用
  const wxsImportNormalizeTokens: Token[] = []
  // 移除内联 wxs 的 lang 属性
  const removeWxsLangAttrTokens: Token[] = []
  // 事件转义
  const eventTokens: Token[] = []
  // tag 调用栈
  const tagStack: string[] = []
  const parser = new Parser(
    {
      onopentagname(name) {
        tagStack.push(name)
        currentTagName = name
        importAttrs = srcImportTagsMap[currentTagName]
        tagStartIndex = parser.startIndex
      },
      onattribute(name, value, quote) {
        attrs[name] = value
        if (importAttrs && currentTagName) {
          for (const attrName of importAttrs) {
            if (attrName === name) {
              deps.push({
                name,
                value,
                quote,
                tagName: currentTagName,
                start: parser.startIndex,
                end: parser.endIndex,
                attrs,
              })
              if (currentTagName === 'wxs' && name === 'src') {
                if (/\.wxs.[jt]s$/.test(value)) {
                  // 5 是 'src="'.length
                  // 1 是 '"'.length
                  wxsImportNormalizeTokens.push(
                    {
                      start: parser.startIndex + 5,
                      end: parser.endIndex - 1,
                      value,
                    },
                  )
                }
              }
            }
          }
        }
        // 事件绑定
        if (name.startsWith('@')) {
          const start = parser.startIndex
          const end = parser.startIndex + name.length
          const { dir, mods } = name.split('.')
            .filter(x => x)
            .reduce<{ dir: string, mods: Record<string, boolean> }>((acc, cur, idx) => {
              if (idx === 0) {
                acc.dir = cur
              }
              else {
                acc.mods[cur] = true
              }
              return acc
            }, { dir: '', mods: {} })

          let rep: string
          if (mods.catch && mods.capture) {
            rep = `capture-catch:${dir.slice(1)}`
          }
          else if (mods.catch) {
            rep = `catch:${dir.slice(1)}`
          }
          else if (mods.mut) {
            rep = `mut-bind:${dir.slice(1)}`
          }
          else if (mods.capture) {
            rep = `capture-bind:${dir.slice(1)}`
          }
          else {
            rep = `bind:${dir.slice(1)}`
          }
          if (rep) {
            eventTokens.push({
              start,
              end,
              value: rep,
            })
          }
        }
        // 移除内联 wxs 的 lang
        if (currentTagName === 'wxs' && name === 'lang' && jsExtensions.includes(value)) {
          removeWxsLangAttrTokens.push({
            start: parser.startIndex,
            end: parser.endIndex,
            value,
          })
        }
      },
      onclosetag() {
        currentTagName = tagStack.pop()
        if (currentTagName && !opts.excludeComponent(currentTagName)) {
          if (Array.isArray(components[currentTagName])) {
            components[currentTagName].push({
              start: tagStartIndex,
              end: parser.endIndex + 1,
            })
          }
          else {
            components[currentTagName] = [{
              start: tagStartIndex,
              end: parser.endIndex + 1,
            }]
          }
        }

        currentTagName = ''
        attrs = {}
        importAttrs = undefined
        tagStartIndex = 0
      },
      ontext(data) {
        if (currentTagName === 'wxs' && jsExtensions.includes(attrs.lang)) {
          inlineWxsTokens.push({
            start: parser.startIndex,
            end: parser.endIndex,
            value: data,
          })
        }
      },
      // <!--  #ifdef  %PLATFORM% -->
      // 平台特有的组件
      // <!--  #endif -->
      oncomment(data) {
        let match = /#ifdef\s+(\w+)/.exec(data)
        if (match) {
          if (match[1] !== opts.platform) {
            removeStartStack.push(parser.startIndex)
          }
        }
        match = /#endif/.exec(data)
        if (match) {
          removeEndStack.push(parser.endIndex + 1)
        }
        commentTokens.push({
          start: parser.startIndex,
          end: parser.endIndex + 1,
          value: data,
        })
      },
    },
    {
      lowerCaseTags: false,
      xmlMode: true,
    },
  )
  parser.write(
    ms,
  )
  parser.end()

  const token: WxmlToken = {
    components,
    deps,
    removeStartStack,
    removeEndStack,
    commentTokens,
    inlineWxsTokens,
    wxsImportNormalizeTokens,
    removeWxsLangAttrTokens,
    eventTokens,
    code: ms,
  }
  scanWxmlCache.set(cacheKey, token)
  return token
}

export type ScanWxmlResult = ReturnType<typeof scanWxml>
