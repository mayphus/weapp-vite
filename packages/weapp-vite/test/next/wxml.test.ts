import { processWxml } from '@/wxml'

describe('wxml', () => {
  it('processWxml remove wxs', () => {
    const { deps, code } = processWxml(`<!-- index.wxml -->
<template name="staffName">
  <view>FirstName: {{firstName}}, LastName: {{lastName}}</view>
</template>
<navigation-bar></navigation-bar>
<image src="../../assets/logo.png" class="w-40 h-40"></image>
<image src="{{logoUrl}}" class="w-40 h-40"></image>
<ice-avatar></ice-avatar>
<scroll-view scroll-y type="list">
  <view class="p-2 border m-2 rounded">{{ motto }}</view>
  <view class="bg-[#c50d69] text-[10px] text-[#123456] w-[323px] {{true?'h-[30px]':'h-[45px]'}}">
    bg-[#7d7ac2] text-[10px] text-[#123456] w-[323px]
  </view>
  <view class="p-2 border">{{message}}</view>
  <view wx:for="{{array}}">{{item}}</view>
  <template is="staffName" data="{{...staffA}}"></template>
  <template is="staffName" data="{{...staffB}}"></template>
  <template is="staffName" data="{{...staffC}}"></template>
  <template bindtap="switchTemplate" is="{{flag? 'odd': 'staffName'}}" data="{{...staffC}}"></template>
  <t-divider></t-divider>
  <view class="user-motto text-pink-200 bg-slate-100">{{m1.message}}</view>
  <wxs module="wxs" src="./test.wxs"></wxs>
  <view id="tapTest" data-hi="Weixin" bindtap="{{wxs.tapName}}">Click me!</view>
</scroll-view>
<template name="odd">
  <view>odd</view>
</template>
<template name="even">
  <view>even</view>
</template>
<view class="button-example">
  <t-button bind:tap="switchTemplate" theme="primary" size="large">填充按钮</t-button>
  <t-button bind:click="switchTemplate" theme="light" size="large">填充按钮</t-button>
  <t-button size="large">填充按钮</t-button>
</view>
<block wx:for="{{[1, 2, 3, 4, 5]}}">
  <template is="{{item % 2 == 0 ? 'even' : 'odd'}}" />
</block>
<wxs src="./index.wxs" module="tools" />
<wxs module="m1">
var msg = "hello world";

module.exports.message = msg;
</wxs>`)
    expect(deps).matchSnapshot('deps')
    expect(code).matchSnapshot('code')
  })

  it('processWxml remove wxs case 0', () => {
    const { deps, code } = processWxml(`<wxs src="./index.wxs" module="tools" />`)
    expect(deps).matchSnapshot('deps')
    expect(code).matchSnapshot('code')
  })

  it('processWxml remove wxs case 1', () => {
    const { deps, code } = processWxml(`<wxs src="./index.wxs.ts" module="tools" />`)
    expect(deps).matchSnapshot('deps')
    expect(code).matchSnapshot('code')
  })

  it('processWxml remove wxs case 2', () => {
    const { deps, code } = processWxml(`<wxs src="./index.wxs.js" module="tools" />`)
    expect(deps).matchSnapshot('deps')
    expect(code).matchSnapshot('code')
  })

  it('processWxml remove wxs case 3', () => {
    const { deps, code } = processWxml(`<wxs module="foo">
var some_msg = "hello world";
module.exports = {
  msg : some_msg,
}
</wxs>
<view> {{foo.msg}} </view>`)
    expect(deps).matchSnapshot('deps')
    expect(code).matchSnapshot('code')
  })

  it('processWxml remove wxs case 4', () => {
    const { deps, code } = processWxml(`<wxs module="foo" lang="js">
var some_msg = "hello world";
module.exports = {
  msg : some_msg,
}
</wxs>
<view> {{foo.msg}} </view>`)
    expect(deps).matchSnapshot('deps')
    expect(code).matchSnapshot('code')
  })

  it('processWxml remove wxs case 5', () => {
    const { deps, code } = processWxml(`<wxs module="foo" lang="ts">
var some_msg = "hello world";
module.exports = {
  msg : some_msg,
}
</wxs>
<view> {{foo.msg}} </view>`)
    expect(deps).matchSnapshot('deps')
    expect(code).matchSnapshot('code')
  })
})
