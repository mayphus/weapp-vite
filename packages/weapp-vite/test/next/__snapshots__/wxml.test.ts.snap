// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`wxml > processWxml remove wxs > code 1`] = `
"<!-- index.wxml -->
<template name="staffName">
  <view>FirstName: {{firstName}}, LastName: {{lastName}}</view>
</template>
<navigation-bar></navigation-bar>
<image src="../../assets/logo.png" class="w-40 h-40"></image>
<image src="{{logoUrl}}" class="w-40 h-40"></image>
<ice-avatar></ice-avatar>
<scroll-view scroll-y type="list">
  <view class="p-2 border m-2 rounded">{{ motto }}</view>
  <view class="bg-[#c50d69] text-[10px] text-[#123456] w-[323px] {{true?'h-[30px]':'h-[45px]'}}">
    bg-[#7d7ac2] text-[10px] text-[#123456] w-[323px]
  </view>
  <view class="p-2 border">{{message}}</view>
  <view wx:for="{{array}}">{{item}}</view>
  <template is="staffName" data="{{...staffA}}"></template>
  <template is="staffName" data="{{...staffB}}"></template>
  <template is="staffName" data="{{...staffC}}"></template>
  <template bindtap="switchTemplate" is="{{flag? 'odd': 'staffName'}}" data="{{...staffC}}"></template>
  <t-divider></t-divider>
  <view class="user-motto text-pink-200 bg-slate-100">{{m1.message}}</view>
  <wxs module="wxs" src="./test.wxs"></wxs>
  <view id="tapTest" data-hi="Weixin" bindtap="{{wxs.tapName}}">Click me!</view>
</scroll-view>
<template name="odd">
  <view>odd</view>
</template>
<template name="even">
  <view>even</view>
</template>
<view class="button-example">
  <t-button bind:tap="switchTemplate" theme="primary" size="large">填充按钮</t-button>
  <t-button bind:click="switchTemplate" theme="light" size="large">填充按钮</t-button>
  <t-button size="large">填充按钮</t-button>
</view>
<block wx:for="{{[1, 2, 3, 4, 5]}}">
  <template is="{{item % 2 == 0 ? 'even' : 'odd'}}" />
</block>
<wxs src="./index.wxs" module="tools" />
<wxs module="m1">
var msg = "hello world";

module.exports.message = msg;
</wxs>"
`;

exports[`wxml > processWxml remove wxs > deps 1`] = `
[
  {
    "attrs": {
      "module": "wxs",
      "src": "./test.wxs",
    },
    "end": 1064,
    "name": "src",
    "quote": """,
    "start": 1048,
    "tagName": "wxs",
    "value": "./test.wxs",
  },
  {
    "attrs": {
      "module": "tools",
      "src": "./index.wxs",
    },
    "end": 1644,
    "name": "src",
    "quote": """,
    "start": 1627,
    "tagName": "wxs",
    "value": "./index.wxs",
  },
]
`;

exports[`wxml > processWxml remove wxs case 0 > code 1`] = `"<wxs src="./index.wxs" module="tools"  />"`;

exports[`wxml > processWxml remove wxs case 0 > deps 1`] = `
[
  {
    "attrs": {
      "lang": "ts",
      "module": "tools",
      "src": "./index.wxs",
    },
    "end": 22,
    "name": "src",
    "quote": """,
    "start": 5,
    "tagName": "wxs",
    "value": "./index.wxs",
  },
]
`;

exports[`wxml > processWxml remove wxs case 1 > code 1`] = `"<wxs src="./index.wxs" module="tools"  />"`;

exports[`wxml > processWxml remove wxs case 1 > deps 1`] = `
[
  {
    "attrs": {
      "lang": "js",
      "module": "tools",
      "src": "./index.wxs.ts",
    },
    "end": 25,
    "name": "src",
    "quote": """,
    "start": 5,
    "tagName": "wxs",
    "value": "./index.wxs.ts",
  },
]
`;

exports[`wxml > processWxml remove wxs case 2 > code 1`] = `"<wxs src="./index.wxs" module="tools" />"`;

exports[`wxml > processWxml remove wxs case 2 > deps 1`] = `
[
  {
    "attrs": {
      "module": "tools",
      "src": "./index.wxs.js",
    },
    "end": 25,
    "name": "src",
    "quote": """,
    "start": 5,
    "tagName": "wxs",
    "value": "./index.wxs.js",
  },
]
`;

exports[`wxml > processWxml remove wxs case 3 > code 1`] = `
"<wxs module="foo">
var some_msg = "hello world";
module.exports = {
  msg : some_msg,
}
</wxs>
<view> {{foo.msg}} </view>"
`;

exports[`wxml > processWxml remove wxs case 3 > deps 1`] = `[]`;

exports[`wxml > processWxml remove wxs case 4 > code 1`] = `
"<wxs module="foo" >
var some_msg = "hello world";
module.exports = {
  msg: some_msg
};
</wxs>
<view> {{foo.msg}} </view>"
`;

exports[`wxml > processWxml remove wxs case 4 > deps 1`] = `[]`;

exports[`wxml > processWxml remove wxs case 5 > code 1`] = `
"<wxs module="foo" >
var some_msg = "hello world";
module.exports = {
  msg: some_msg
};
</wxs>
<view> {{foo.msg}} </view>"
`;

exports[`wxml > processWxml remove wxs case 5 > deps 1`] = `[]`;
